file_transfer/src/file_transfer.c

```diff
@@ -0,0 +1,319 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * \example file_transfer.c
+ *
+ * File Transfer Example.
+ *
+ * Demonstrates how to use the PolySync file transfer API to initiate and
+ * support file transfer operations between nodes.
+ *
+ */
+
+
+
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <signal.h>
+#include <semaphore.h>
+
+// PolySync API headers
+#include "polysync_core.h"
+#include "polysync_sdf.h"
+#include "polysync_message.h"
+#include "polysync_node.h"
+
+// Header for PolySync File Transfer API,
+// which is the topic of this example.
+#include "polysync_file_transfer.h"
+
+
+// All polysync-manager nodes support file transfer operations.
+// This example will attempt a file transfer with the first polysync-manager
+// it can find. This semaphore is used to signal when the first polysync-manager
+// has responded, and its GUID has been captured.
+sem_t manager_response;
+
+
+// After a polysync-manager node has been requested to provide its status,
+// it will respond. This is the name of the type of that message.
+const char manager_status_msg_name[] = "ps_response_msg";
+
+
+// When polysync-manager responds with its status,
+// this function will capture its GUID and signal 
+// that the program may proceeed via the semaphore.
+static void ps_manager_status_msg_handler(
+        const ps_msg_type msg_type,
+        const ps_msg_ref const message,
+        void * const user_data )
+{
+    // The pointer to the GUID of the first responding polysync-manager node
+    // was passed as the user_data field.
+    ps_guid * const manager_guid = (ps_guid * const)user_data;
+
+    // This function is only subscribed to 'ps_manager_status_msg' and
+    // so it is safe to cast the message reference to that type.
+    const ps_response_msg * const manager_status =
+            (ps_response_msg*)message;
+
+    // Set the manager GUID based on the header of the response.
+    *manager_guid = manager_status->header.src_guid;
+
+    // Signal to the main thread to resume because
+    // the manager GUID has been retrieved.
+    sem_post( &manager_response );
+}
+
+
+// This function allocates and publishes a command
+// for all polysync-manager nodes to provide their statuses.
+static int request_manager_response(
+        const ps_node_ref node_ref )
+{
+    const char command_msg_name[] = "ps_command_msg";
+    ps_msg_type command_msg_type = PSYNC_MSG_TYPE_INVALID;
+
+    int ret;
+
+    ps_command_msg * const command_msg;
+
+    if( (ret = psync_message_get_type_by_name(
+            node_ref,
+            command_msg_name,
+            &command_msg_type )) != DTC_NONE )
+    {
+        printf( "failed to get message type of %s\n", command_msg_name );
```

Comment:
any reason not to use the PolySync log macros so these messages end up in the polysync.log?